### PR TITLE
Spec Header Dirs: Only first include/

### LINKED ISSUE #13969: cuda: g++ finds C++ std headers shipped by cuda sdk instead of its own libstdc++
While trying to build *standard C++ sources (no device code)* with Spack's own `g++` wrapper from within a package's `build-env` that depends on `cuda`, C++ standard headers (e.g.: `vector`) vendored by the `cuda` SDK are found by `g++` before its own `libstdc++`, breaking any trivial compilation that involves C++ standard headers. 

### Steps to reproduce the issue

Any package depending on `cuda` fits the purpose, I've been able to reproduce the issue with a bogus `package.py` like this:

```python
from spack import *

class Nvtest(Package):
    version('0')
    depends_on('cuda')
```

The `test.cpp` source I'm trying to compile uses just one standard header and is as simple as possible:

```cpp
#include <vector>

int main() {
    std::vector<int> v;
}
```

Actual steps:

```console
$ spack install gcc@8.3.0
$ spack compiler add $(spack location -i gcc@8.3.0)
$ spack install cuda%gcc@8.3.0
$ spack build-env nvtest%gcc@8.3.0 -- g++ test.cpp
In file included from /spack/opt/spack/linux-centos7-power8le/gcc-8.3.0/cuda-10.2.89-zamew7asat3psayj22opj5xnm4e
o6j6y/include/cuda/std/detail/libcxx/include/stdlib.h:93,
                 from /spack/opt/spack/linux-centos7-power8le/gcc-8.3.0/cuda-10.2.89-zamew7asat3psayj22opj5xnm4e
o6j6y/include/cuda/std/detail/libcxx/include/cstdlib:86,
                 from /spack/opt/spack/linux-centos7-power8le/gcc-8.3.0/cuda-10.2.89-zamew7asat3psayj22opj5xnm4e
o6j6y/include/cuda/std/detail/libcxx/include/exception:81,
                 from /spack/opt/spack/linux-centos7-power8le/gcc-8.3.0/cuda-10.2.89-zamew7asat3psayj22opj5xnm4e
o6j6y/include/cuda/std/detail/libcxx/include/typeinfo:60,
                 from /spack/opt/spack/linux-centos7-power8le/gcc-8.3.0/cuda-10.2.89-zamew7asat3psayj22opj5xnm4e
o6j6y/include/cuda/std/detail/libcxx/include/memory:653,
                 from /spack/opt/spack/linux-centos7-power8le/gcc-8.3.0/cuda-10.2.89-zamew7asat3psayj22opj5xnm4e
o6j6y/include/cuda/std/detail/libcxx/include/algorithm:643,
                 from /spack/opt/spack/linux-centos7-power8le/gcc-8.3.0/cuda-10.2.89-zamew7asat3psayj22opj5xnm4e
o6j6y/include/cuda/std/detail/libcxx/include/__bit_reference:15,
                 from /spack/opt/spack/linux-centos7-power8le/gcc-8.3.0/cuda-10.2.89-zamew7asat3psayj22opj5xnm4e
o6j6y/include/cuda/std/detail/libcxx/include/vector:274,
                 from main.cpp:1:
/spack/opt/spack/linux-centos7-power8le/gcc-4.8.5/gcc-8.3.0-kvs4zwy3ojuvmc43h4yf364ie2zy5gtk/include/c++/8.3.0/s
tdlib.h:38:12: error: 'std::abort' has not been declared
 using std::abort;
# lots of similar errors omitted
```

The same behaviour is observed for both `gcc@9.2.0` (**not supported** by `cuda@10.2.89`) and system-shipped `gcc@4.8.5` (**supported** by `cuda@10.2.89`).

I think that the issue stems from this:

```console
$ spack build-env nvtest%gcc@8.3.0 | grep SPACK_INCLUDE_DIRS
SPACK_INCLUDE_DIRS=/spack/opt/spack/linux-centos7-power8le/gcc-8.3.0/cuda-10.2.89-zamew7asat3psayj22opj5xnm4eo6j6y/include:/spack/opt/spack/linux-centos7-power8le/gcc-8.3.0/cuda-10.2.89-zamew7asat3psayj22opj5xnm4eo6j6y/include/cuda/std/detail/libcxx/include
```

The frontend confirms the inclusion search order:

```console
$ spack build-env nvtest%gcc@8.3.0 -- g++ -E -v -
Reading specs from /spack/opt/spack/linux-centos7-power8le/gcc-4.8.5/gcc-8.3.0-kvs4zwy3ojuvmc43h4yf364ie2zy5gtk/lib/gcc/powerpc64le-unknown-linux-gnu/8.3.0/specs
#include "..." search starts here:
#include <...> search starts here:
 /spack/opt/spack/linux-centos7-power8le/gcc-8.3.0/cuda-10.2.89-zamew7asat3psayj22opj5xnm4eo6j6y/include
 /spack/opt/spack/linux-centos7-power8le/gcc-8.3.0/cuda-10.2.89-zamew7asat3psayj22opj5xnm4eo6j6y/include/cuda/std/detail/libcxx/include
 /spack/opt/spack/linux-centos7-power8le/gcc-4.8.5/gcc-8.3.0-kvs4zwy3ojuvmc43h4yf364ie2zy5gtk/lib/gcc/powerpc64le-unknown-linux-gnu/8.3.0/include
 /usr/local/include
 /spack/opt/spack/linux-centos7-power8le/gcc-4.8.5/gcc-8.3.0-kvs4zwy3ojuvmc43h4yf364ie2zy5gtk/include
 /spack/opt/spack/linux-centos7-power8le/gcc-4.8.5/gcc-8.3.0-kvs4zwy3ojuvmc43h4yf364ie2zy5gtk/lib/gcc/powerpc64le-unknown-linux-gnu/8.3.0/include-fixed
 /usr/include
End of search list.
```

While `<cuda>/include` provides SDK's own headers, `<cuda>/include/cuda/std/detail/libcxx/include` pulls in standard headers provided by a vendored implementation of the C++ standard library. The fact that a path component is called `/detail/` makes me think that it should be treated as an implementation detail (as is usually done in C++ code bases) and not explicitly injected in the inclusion search space.

### PR DESCRIPTION
**Pull request by ax3l**
Avoid matching recurringly nested include paths that most certainly refer to internally shipped libraries in packages.
Example in CUDA Toolkit #13819 #13983, shipping a libc++ fork internally with libcu++ since 10.2.89: `<prefix>/include/cuda/std/detail/libcxx/include`.

Quick fix #13969 

A more thorough solution is not to recursively search for headers anymore since this is against the idea of C/C++ facade headers: https://github.com/spack/spack/issues/13969#issuecomment-561929481 Unfortunately, it's not that easy for us to find which headers are facade and which ones are detail, but I am not sure we have to know anything besides the include dir locations.

Generally, just searching for *any* `include/` in the overall path (instead of just taking the `<prefix>/include`) is dangerous as well. For example, spack could be installed in `/my/large/software/include/here/is/axels/home/src/spack/` ;-)

General policy suggestion: just add `<prefix>/include` to the build environment by default. Packages that break this unix-wide default should provide the
```py
@property
def headers(self):
```

@adamjstewart mentioned on Slack that some Python packages might need special treatment, let's see if we can find examples again to double check.

### CODE REVIEW

**Comment by adamjstewart** [comment]
For the record, `boost` was the package I needed a way to locate the Python headers for. The following packages use the same logic:
```console
$ grep 'spec.*python.*headers' */package.py
boost/package.py:            spec['python'].headers.directories[0],
dealii/package.py:                python_include = spec['python'].headers.directories[0]
neuron/package.py:            py_inc = spec['python'].headers.directories[0]
opencv/package.py:            python_include_dir = spec['python'].headers.directories[0]
openspeedshop-utils/package.py:        python_include = spec['python'].headers.directories[0]
openspeedshop/package.py:        python_include = spec['python'].headers.directories[0]
precice/package.py:            python_include = spec['python'].headers.directories[0]
psi4/package.py:                os.path.join(spec['python'].headers.directories[0]),
py-gpaw/package.py:        python_include = spec['python'].headers.directories[0]
```
I agree that the default should not be recursively searching for headers, we should just use `prefix.include` unless someone overrides `headers`.

**Comment by ax3l** [comment]
So it's actually the CPython headers you are searching for in those packages. We then should just overwrite the `python` package's headers property :)

**Comment by ax3l** [comment]
Hm, CI says no. Have to go back to regex school.

**Comment by alalazo** [comment]
@ax3l https://regex101.com/ :slightly_smiling_face:

**Comment by ax3l** [comment]
Pew the regex was fine, just forgot to update the group number below :D

**Comment by ax3l** [comment]
Ha, just because unit tests pass locally does not mean CI does. Whelp.

**Comment by ax3l** [comment]
Ha, what am I saying. TIL: non-capture groups.

**Comment by nazavode** [comment]
Just tried this patch on top of f84ad57fe7c907d8be90d4beaaba151926d74ec0 and it looks like it is working fine, I was able to build stuff that was previously broken like `openmpi+cuda ^hwloc+cuda ^cuda@10.2.89`. Thanks!

File: lib/spack/llnl/util/filesystem.py [hunk 1]
```diff
@@ -1156,7 +1156,9 @@ class HeaderList(FileList):
 
     # Make sure to only match complete words, otherwise path components such
     # as "xinclude" will cause false matches.
-    include_regex = re.compile(r'(.*)(\binclude\b)(.*)')
+    # Avoid matching paths such as <prefix>/include/something/detail/include,
+    # e.g. in the CUDA Toolkit which ships internal libc++ headers.
+    include_regex = re.compile(r'((?:(?!(?:\binclude\b)).)*)(\binclude\b)(.*)')
```
Comment (alalazo) [inline lib/spack/llnl/util/filesystem.py]:
```suggestion
    include_regex = re.compile(r'(.*?)(\binclude\b)(.*)')
```
Not a regex master (proven by the fact that I don't understand at a first glance what's going on here :laughing: ). By reading just the comment though I wonder if the simpler modifications above works fine. It basically says that the first match should be non-greedy.

Comment (ax3l) [reply lib/spack/llnl/util/filesystem.py]: Wow, yes :D

**Comment by ax3l** [comment]
Just adding a unit test as well before pushing the fix.

**Pull request merged**